linked_module_resolver_test.go

```diff
@@ -131,7 +131,7 @@ func testLinkedModuleResolver(t *testing.T, context spec.G, it spec.S) {
 
 			context("when the destination cannot be scaffolded", func() {
 				it.Before(func() {
-					Expect(os.Mkdir(filepath.Join(layerPath, "sub-dir"), 0400)).To(Succeed())
```

Comment:
tbh: no idea why the test succeeded "most of the time"

Reply:
After "some" debugging: It worked in 2/3 of the cases because the resolved path in the test data has some subfolder.

1 of the 3 links is without subfolders and iterating over a dict in `go` doesn't give a guaranteed order.

`MkDirAll` will not fail for a r/o folder if the folder already exist and no subfolder is needed.

Reply:
Using WriteFile seems different than creating a sub-directory. Do you have the output from when it fails? That would be good to understand why it was flaky and what the right fix might be.

Reply:
> Using WriteFile seems different than creating a sub-directory. 

It is. That is the reason why it works now :) Imho, `WriteFile` is the correct tool for that anyway. Unfortunately, the tests are not that self explanatory and we had to make some assumptions. But in the end all made sense.

> Do you have the output from when it fails? 

You can just look in the pr validation of some other prs, e.g. #688 (it roughly happes 1/3 of the time)

```
install/LinkedModuleResolver/Copy/failure_cases/when_the_destination_cannot_be_scaffolded/returns_an_error
    linked_module_resolver_test.go:217: 
        Expected
            <*fmt.wrapError | 0xc000312640>: 
            failed to copy linked module directory to layer path: failed to copy: destination exists: remove /tmp/another-layer1310829255/sub-dir/module-5: permission denied
            {
                msg: "failed to copy linked module directory to layer path: failed to copy: destination exists: remove /tmp/another-layer1310829255/sub-dir/module-5: permission denied",
                err: <*fmt.wrapError | 0xc000312620>{
                    msg: "failed to copy: destination exists: remove /tmp/another-layer1310829255/sub-dir/module-5: permission denied",
                    err: <*fs.PathError | 0xc0004f50b0>{
                        Op: "remove",
                        Path: "/tmp/another-layer1310829255/sub-dir/module-5",
                        Err: <syscall.Errno>0xd,
                    },
                },
            }
        to match error
            <*matchers.ContainSubstringMatcher | 0xc0004f50e0>: {
                Substr: "failed to setup linked module directory scaffolding",
                Args: nil,
            }
```

> That would be good to understand why it was flaky and what the right fix might be.

We thought so as well, so after fixing it, we had a "little" debugging session to understand *why* the test is flaky. In a nutshell:

* The test creates a r/o directory to provoke a very specific error when a subdirectory is created via `MkDirAll`
* In `go`, iterating over a dict is not ordered, but the order is random
* The test data needs to create links in a subdirectory that cannot be created
* EXCEPT: One module (module-5) that doesn't need a subdirectory
* The call will still fail, but with a different error because the r/o directory should be removed

When creating a file instead of directory, it works

* You cannot create a subdirectory to a file
* `MkDirAll` fails if you want to create a directory with the same name as a file

Reply:
Yes, the test is flakky.
On this part here https://github.com/paketo-buildpacks/npm-install/blob/7b388f01af53beba6500526d3787d119610a50e5/linked_module_resolver_test.go#L134 it sets the permissions.
If the first iteration that happens  here  https://github.com/paketo-buildpacks/npm-install/blob/7b388f01af53beba6500526d3787d119610a50e5/linked_module_resolver.go#L58 the module is anything else except the module-5  https://github.com/paketo-buildpacks/npm-install/blob/7b388f01af53beba6500526d3787d119610a50e5/linked_module_resolver_test.go#L45-L67 it manages to error on the first loop as it is not able to create the directory `/tmp/layer2780743255/sub-dir/src/packages` as the `sub-dir` has permission 0400 , this is what we expect and as a resutl the test passes. But in case of the `module-5` it will try to create the directory `/tmp/layer2780743255/sub-dir` but because of the fact that `sub-dir` directory already exists, the `mkdirAll` function will not fail as it should and will fail on the below if statement https://github.com/paketo-buildpacks/npm-install/blob/7b388f01af53beba6500526d3787d119610a50e5/linked_module_resolver.go#L63  as a result throwing a different error than the expected.

Reply:
So another solution in the current context could be 
```
			context("when the destination cannot be scaffolded", func() {
				it.Before(func() {
					Expect(os.MkdirAll(filepath.Join(layerPath, "sub-dir", "another-sub-dir"), os.ModePerm)).To(Succeed())
					Expect(os.Chmod(filepath.Join(layerPath, "sub-dir"), 0400)).To(Succeed())
				})

				it("returns an error", func() {
					err := resolver.Resolve(filepath.Join(workspace, "package-lock.json"), filepath.Join(layerPath, "sub-dir", "another-sub-dir"))
					Expect(err).To(MatchError(ContainSubstring("failed to setup linked module directory scaffolding")))
				})
                                // Revert the permissions to allow proper clean up
				it.After(func() {
					Expect(os.Chmod(filepath.Join(layerPath, "sub-dir"), 0700)).To(Succeed())
				})
			})

```

Which practically creates a sub-sub-directory with the proper permissions in order for the module-5 to fail.

Reply:
> So another solution in the current context could be

Are you sure that would fix the problem?

https://github.com/paketo-buildpacks/npm-install/blob/7b388f01af53beba6500526d3787d119610a50e5/linked_module_resolver.go#L63

Would still work because the directory already exists. I haven't checked it yet, though.